File: apps/api/internal/pkgcheck/contract_test.go
// The contract test: proves fixtures/format-v2/ and this package still
// agree with packages/course-format/src/contract.test.ts, the TypeScript
// half of the same requirement.
//
// Task 3 of the server-side pivot built fixtures/format-v2/ — real package
// directories on disk, hostile/<case>/ plus valid-course/ — that BOTH
// implementations of the rule set (that TypeScript module, and this Go
// port) are required to run against. A case directory is a MINIMAL package
// (a valid v2 manifest plus exactly one file carrying its violation) and
// one expect.json, shaped {"codes": ["SCRIPT_TAG"]}, that sits OUTSIDE the
// package contents — read directly by this file, but excluded when the
// case directory is zipped into package bytes.
//
// The contract is SUBSET, not equality: expected.codes must be a subset of
// the codes Validate actually returns. The scanner deliberately
// over-reports (see content.go's own header), so a case may legitimately
// trip more rules than the one it was built to name. valid-course is the
// one exception, checked for EQUALITY with the empty set: it is the
// package a later task serves to a real browser, so it must be flawless.
//
// This file covers the CONTENT cases only — script-tag, event-handler-attr,
// javascript-url, embedded-frame, tier-field, loose-js — because Task 7
// adds the eight WIDGET_* rules and owns the remaining five hostile cases
// (widget-extra-file, widget-orphan, widget-external-url, widget-missing,
// widget-forbidden-api). contentCases below is an EXPLICIT slice, not a
// directory listing: globbing every hostile/* directory and silently
// skipping the ones this task does not cover would mean a renamed or
// newly-added case fails to get its own test with no signal that it was
// ever supposed to. wantAllHostileCasesCovered documents (and checks) that
// the two lists partition the whole hostile/ corpus between this task and
// Task 7, so a case neither list names fails loudly instead of quietly
// being skipped by both.
package pkgcheck

import (
	"archive/zip"
	"bytes"
	"encoding/json"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"testing"
)

const fixturesRoot = "../../../../fixtures/format-v2"

const expectFile = "expect.json"

// contentCases is the explicit, non-glob list this task is responsible
// for. See the file comment for why this may not become a directory scan.
var contentCases = []string{
	"script-tag",
	"event-handler-attr",
	"javascript-url",
	"embedded-frame",
	"tier-field",
	"loose-js",
}

// widgetCases is Task 7's five, named here ONLY so
// TestHostileCorpusIsFullyAccountedFor can prove contentCases plus
// widgetCases covers every directory under hostile/ — not to run them:
// running them is that task's job, against rules this package does not
// implement yet.
var widgetCases = []string{
	"widget-extra-file",
	"widget-orphan",
	"widget-external-url",
	"widget-missing",
	"widget-forbidden-api",
}

type expectedFindings struct {
	Codes []string `json:"codes"`
}

// readExpect reads and validates one case's expect.json. Fails the test
// loudly (never skips) when the file is missing or malformed — an
// undeclared case is a corpus defect, not a reason to say nothing.
func readExpect(t *testing.T, dir string) expectedFindings {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join(dir, expectFile))
	if err != nil {
		t.Fatalf("%s: missing %s — every hostile case MUST declare its expected codes (e.g. {\"codes\": [\"SCRIPT_TAG\"]}); a case without this file must not be silently skipped: %v", dir, expectFile, err)
	}
	var got expectedFindings
	if err := json.Unmarshal(raw, &got); err != nil {
		t.Fatalf("%s/%s: invalid JSON: %v", dir, expectFile, err)
	}
	if len(got.Codes) == 0 {
		t.Fatalf("%s/%s: \"codes\" must be a non-empty array", dir, expectFile)
	}
	return got
}

// zipDir zips every file under dir into package bytes, package-relative
// paths using "/" regardless of host OS, EXCLUDING expect.json wherever it
// sits — the one file a case directory keeps outside its own package
// contents. This round-trips through a real archive/zip.Writer, the same
// door a real package arrives by, rather than building a files map by
// hand and skipping the zip layer.
func zipDir(t *testing.T, dir string) []byte {
	t.Helper()

	var buf bytes.Buffer
	zw := zip.NewWriter(&buf)

	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		if d.Name() == expectFile {
			return nil
		}
		rel, err := filepath.Rel(dir, path)
		if err != nil {
			return err
		}
		rel = filepath.ToSlash(rel)

		data, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		w, err := zw.Create(rel)
		if err != nil {
			return err
		}
		_, err = w.Write(data)
		return err
	})
	if err != nil {
		t.Fatalf("zipDir(%s): %v", dir, err)
	}
	if err := zw.Close(); err != nil {
		t.Fatalf("zipDir(%s): closing zip writer: %v", dir, err)
	}
	return buf.Bytes()
}

// codeSet runs Validate over dir's zipped contents and returns the set of
// codes it reported.
func codeSet(t *testing.T, dir string) map[string]bool {
	t.Helper()
	findings, _, err := Validate(zipDir(t, dir))
	if err != nil {
		t.Fatalf("Validate(%s): unexpected system error: %v", dir, err)
	}
	set := make(map[string]bool, len(findings))
	for _, f := range findings {
		set[f.Code] = true
	}
	return set
}

func TestHostileContentCases(t *testing.T) {
	if len(contentCases) == 0 {
		// If this ever loses its entries, every case below is a t.Run
		// that never runs — a describe-block-shaped false green, exactly
		// what contract.test.ts's "kho hostile khong rong" guards against
		// on the TypeScript side. Keep the analogous guard here too.
		t.Fatal("contentCases is empty — this test would silently pass without checking anything")
	}

	for _, name := range contentCases {
		t.Run(name, func(t *testing.T) {
			dir := filepath.Join(fixturesRoot, "hostile", name)
			if _, err := os.Stat(dir); err != nil {
				t.Fatalf("listed in contentCases but not found on disk: %v", err)
			}

			expected := readExpect(t, dir)
			codes := codeSet(t, dir)

			var missing []string
			for _, code := range expected.Codes {
				if !codes[code] {
					missing = append(missing, code)
				}
			}
			if len(missing) > 0 {
				var got []string
				for c := range codes {
					got = append(got, c)
				}
				sort.Strings(got)
				t.Errorf("hostile/%s: missing expected code(s) %v in findings — got %v", name, missing, got)
			}
		})
	}
}

func TestValidCourseHasZeroFindings(t *testing.T) {
	findings, pkg, err := Validate(zipDir(t, filepath.Join(fixturesRoot, "valid-course")))
	if err != nil {
		t.Fatalf("unexpected system error: %v", err)
	}
	if len(findings) != 0 {
		t.Fatalf("valid-course must be perfectly clean, but got %d finding(s): %+v", len(findings), findings)
	}
	// findings empty implies pkg != nil — Validate's own documented
	// invariant, checked here rather than assumed.
	if pkg == nil {
		t.Fatal("Validate returned zero findings but a nil *Package — violates its own documented invariant")
	}
	if len(pkg.Chapters) != 2 {
		t.Errorf("valid-course has 2 chapters in its manifest, got %d", len(pkg.Chapters))
	}
	if _, ok := pkg.Widgets["dem-so"]; !ok {
		t.Errorf("valid-course ships widgets/dem-so/index.html, but Package.Widgets has no \"dem-so\" entry (got keys %v)", widgetKeys(pkg.Widgets))
	}
}

func widgetKeys(m map[string]string) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

// TestHostileCorpusIsFullyAccountedFor proves contentCases and widgetCases
// together name every directory under hostile/ — neither more (a stale
// name for a deleted case) nor fewer (a new case neither list has caught
// up with, which would otherwise run under neither task's test file).
func TestHostileCorpusIsFullyAccountedFor(t *testing.T) {
	entries, err := os.ReadDir(filepath.Join(fixturesRoot, "hostile"))
	if err != nil {
		t.Fatalf("reading hostile/: %v", err)
	}

	onDisk := make(map[string]bool, len(entries))
	for _, e := range entries {
		if e.IsDir() {
			onDisk[e.Name()] = true
		}
	}
	if len(onDisk) == 0 {
		t.Fatal("fixtures/format-v2/hostile/ has no case directories at all")
	}

	accounted := make(map[string]bool, len(contentCases)+len(widgetCases))
	for _, name := range contentCases {
		accounted[name] = true
	}
	for _, name := range widgetCases {
		accounted[name] = true
	}

	for name := range onDisk {
		if !accounted[name] {
			t.Errorf("hostile/%s exists on disk but is in neither contentCases (this task) nor widgetCases (Task 7) — it would silently run under no test", name)
		}
	}
	for name := range accounted {
		if !onDisk[name] {
			t.Errorf("%q is listed in contentCases/widgetCases but has no hostile/%s directory on disk", name, name)
		}
	}
}

File: packages/course-format/src/contract.test.ts
/**
 * The contract test: proves `fixtures/format-v2/` and this module still agree.
 *
 * Task 3 of the server-side pivot builds a corpus of real package directories
 * on disk — `fixtures/format-v2/hostile/<case>/` plus `valid-course/` — that
 * BOTH implementations of the rule set (this TypeScript one, and the Go port
 * `apps/api/internal/pkgcheck` from tasks 6–7) are required to run against.
 * This file is the TypeScript half of that requirement. Its job is to iterate
 * the corpus, not to know anything about individual cases — a case's own
 * `expect.json` is where knowledge about THAT case lives, so the Go suite can
 * read the identical file and this test never has to be edited when a case is
 * added, only when the corpus itself changes shape.
 *
 * Each hostile case directory is a MINIMAL package (a valid v2 manifest plus
 * exactly one file carrying its violation) plus one `expect.json`, shaped
 * `{"codes": ["SCRIPT_TAG"]}`, that sits OUTSIDE the package contents — a
 * case dir is zipped for real, through this package's own `packZip`/
 * `unpackZip`, with `expect.json` excluded, and the resulting bytes are handed
 * to `validatePackage` exactly as a consumer would receive them.
 *
 * The contract is SUBSET, not equality: `expected.codes ⊆ codes(findings)`.
 * The scan deliberately over-reports (see `validate.ts`'s own header), so a
 * case may legitimately trip more rules than the one it was built to name.
 * `valid-course` is the one exception and is checked for EQUALITY with the
 * empty set — it is the package Task 16 serves to a real browser, so it must
 * be flawless, not just "flawless enough for this one rule".
 *
 * Two properties this file exists to guarantee, both load-bearing:
 *  - A case directory with no `expect.json` FAILS loudly instead of being
 *    skipped. The corpus is discovered with `readdirSync`, not a hard-coded
 *    list, so a directory added without its `expect.json` still gets its own
 *    `it(...)` — one that throws instead of quietly doing nothing.
 *  - Every failure names the case. Each generated test's title is
 *    `hostile/<case>`, and the per-code assertion message repeats the case
 *    name and the codes that WERE found — a person diagnosing a disagreement
 *    between this file and the Go port needs the case name first, not last.
 *
 * Paths are relative to the process's CWD, same convention `zip.test.ts`
 * already uses for `fixtures/courses/so-dau-phay-dong` (its `ROOT` constant) —
 * `make test-format` and `bun run test` both run with CWD at
 * `packages/course-format`, two directories below the repo root. This is a
 * second copy of that same convention, not a new one: `node-test-env.d.ts`
 * documents why this package hand-declares only the `node:fs` members its
 * tests use rather than pulling in `"types": ["node"]` wholesale, and reaching
 * for `import.meta.url`/`node:path`/`node:url` here would be exactly the
 * "reaching for anything wider" that file exists to make fail first.
 */

import { readFileSync, readdirSync } from 'node:fs';
import { describe, expect, it } from 'vitest';

import { packZip, unpackZip } from './zip';
import { validatePackage } from './validate';

const FORMAT_V2_ROOT = '../../fixtures/format-v2';
const HOSTILE_ROOT = `${FORMAT_V2_ROOT}/hostile`;
const VALID_COURSE_DIR = `${FORMAT_V2_ROOT}/valid-course`;

const EXPECT_FILE = 'expect.json';

interface ExpectedFindings {
  readonly codes: readonly string[];
}

const utf8 = new TextDecoder('utf-8');

/**
 * Reads every file under `dir` into a package-relative `Map`, the shape
 * `packZip`/`validatePackage` take. `expect.json` — the ONE file a hostile
 * case keeps outside its own package contents — is left out wherever it sits,
 * so a case directory zips exactly what a real course package would contain.
 */
function readDirAsPackage(dir: string): Map<string, Uint8Array> {
  const files = new Map<string, Uint8Array>();

  const walk = (abs: string, prefix: string): void => {
    for (const entry of readdirSync(abs, { withFileTypes: true })) {
      const rel = prefix === '' ? entry.name : `${prefix}/${entry.name}`;
      const absChild = `${abs}/${entry.name}`;
      if (entry.isDirectory()) {
        walk(absChild, rel);
        continue;
      }
      if (entry.name === EXPECT_FILE) continue;
      files.set(rel, readFileSync(absChild));
    }
  };

  walk(dir, '');
  return files;
}

/** Zips `dir` through this package's OWN writer, then reads it back through its OWN reader — the exact round trip a real consumer makes, not a shortcut past it. */
function packAndValidate(dir: string): readonly string[] {
  const files = readDirAsPackage(dir);
  const zipped = packZip(files);
  const unpacked = unpackZip(zipped);
  return validatePackage(unpacked).findings.map((f) => f.code);
}

// Discovered, not hard-coded: a case directory that exists on disk gets a test
// whether or not anyone remembered to list it here. This is what makes "a new
// case with no expect.json fails loudly" true — there is no list to forget to
// update.
const hostileCaseNames = readdirSync(HOSTILE_ROOT, { withFileTypes: true })
  .filter((e) => e.isDirectory())
  .map((e) => e.name)
  .sort();

describe('hợp đồng fixtures/format-v2/hostile/*', () => {
  it('kho hostile không rỗng', () => {
    // Nếu ai đó lỡ xoá sạch thư mục hostile, vòng lặp dưới đây sinh ra ĐÚNG
    // KHÔNG `it()` nào — và một describe rỗng là XANH theo vitest. Ca này là
    // thứ duy nhất còn đứng ra nói to hộ điều đó.
    expect(hostileCaseNames.length).toBeGreaterThan(0);
  });

  for (const caseName of hostileCaseNames) {
    it(`hostile/${caseName}: expected.codes ⊆ codes(findings)`, () => {
      const dir = `${HOSTILE_ROOT}/${caseName}`;
      const expectPath = `${dir}/${EXPECT_FILE}`;

      let expectedRaw: string;
      try {
        expectedRaw = utf8.decode(readFileSync(expectPath));
      } catch {
        throw new Error(
          `hostile/${caseName}: thiếu ${EXPECT_FILE} — mỗi case hostile PHẢI khai báo mã mong đợi ` +
            `(vd. {"codes": ["SCRIPT_TAG"]}); một case không có tệp này không được bỏ qua âm thầm.`,
        );
      }
      const expected = JSON.parse(expectedRaw) as ExpectedFindings;
      expect(
        Array.isArray(expected.codes) && expected.codes.length > 0,
        `hostile/${caseName}: ${EXPECT_FILE} phải có trường "codes" là mảng khác rỗng`,
      ).toBe(true);

      const codes = new Set(packAndValidate(dir));
      for (const code of expected.codes) {
        expect(
          codes.has(code),
          `hostile/${caseName}: thiếu mã "${code}" trong findings — có: [${[...codes].sort().join(', ')}]`,
        ).toBe(true);
      }
    });
  }
});

describe('hợp đồng fixtures/format-v2/valid-course', () => {
  it('không có phát hiện nào — đây là gói Task 16 phục vụ cho một trình duyệt thật', () => {
    const codes = packAndValidate(VALID_COURSE_DIR);
    expect(codes, `valid-course phải sạch tuyệt đối, nhưng có: [${codes.join(', ')}]`).toEqual([]);
  });
});
